### PR TITLE
IPv{4,6,Any}{Interface,Network} types added

### PR DESCRIPTION
**Pull request by pilosus**
## Change Summary

As per discussion in #333 support for IP interface and network added

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `HISTORY.rst` has been updated
  * if this is the first change since a release, please add a new section
  * include the issue number or this pull request number `#<number>`
  * include your github username `@<whomever>`

### CODE REVIEW

File: pydantic/types.py [hunk 1]
```diff
@@ -516,3 +528,41 @@ def validate(cls, value: Union[str, bytes, int]) -> Union[IPv4Address, IPv6Addre
...
+            return IPv4Interface(value)
+        except ValueError:
+            pass
+
+        with change_exception(errors.IPvAnyInterfaceError, ValueError):
+            return IPv6Interface(value)
+
+
+class IPvAnyNetwork(_BaseNetwork):  # type: ignore
+    field: str = 'ip_vany_network_strict'
```
Comment (samuelcolvin) [inline pydantic/types.py]: I don't think we can implement `strict` like this, it's unlike anything else we do and will cause confusion.

I think that the default (`strict=True`) is fine. If people really want `strict=False` they can implement `IPvAnyNetworkLax` themselves.

Comment (pilosus) [reply pydantic/types.py]: I've just fixed the ``strict`` argument

File: pydantic/types.py [hunk 2]
```diff
@@ -516,3 +528,37 @@ def validate(cls, value: Union[str, bytes, int]) -> Union[IPv4Address, IPv6Addre
...
             return IPv6Address(value)
+
+
+class IPvAnyInterface(_BaseAddress):
+    @classmethod
+    def __get_validators__(cls) -> 'CallableGenerator':
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, value: Union[str, bytes, int, NetworkType]) -> Union[IPv4Interface, IPv6Interface]:
```
Comment (samuelcolvin) [inline pydantic/types.py]:
```suggestion
    def validate(cls, value: NetworkType) -> Union[IPv4Interface, IPv6Interface]:
```

Or am I missing something?

File: pydantic/types.py [hunk 3]
```diff
@@ -516,3 +528,37 @@ def validate(cls, value: Union[str, bytes, int]) -> Union[IPv4Address, IPv6Addre
...
+            return IPv6Interface(value)
+
+
+class IPvAnyNetwork(_BaseNetwork):  # type: ignore
+    @classmethod
+    def __get_validators__(cls) -> 'CallableGenerator':
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, value: Union[str, bytes, int, NetworkType]) -> Union[IPv4Network, IPv6Network]:
```
Comment (samuelcolvin) [inline pydantic/types.py]:
```suggestion
    def validate(cls, value: NetworkType) -> Union[IPv4Network, IPv6Network]:
```

File: pydantic/validators.py [hunk 4]
```diff
@@ -235,6 +235,51 @@ def ip_v6_address_validator(v: Any) -> IPv6Address:
...
 
+def ip_v4_network_validator(v: Any, values: Dict[str, Any]) -> IPv4Network:
+    """
+    Use ``ip_v4_network_strict`` field for initialisation of IPv4Network
+    with explicit ``strict`` argument.
+
+    See more:
+    https://docs.python.org/library/ipaddress.html#ipaddress.IPv4Network
+    """
+    field: str = 'ip_v4_network_strict'
```
Comment (samuelcolvin) [inline pydantic/validators.py]: As discussed, this should be removed.

File: pydantic/validators.py [hunk 5]
```diff
@@ -235,6 +235,51 @@ def ip_v6_address_validator(v: Any) -> IPv6Address:
...
+
+    if isinstance(v, IPv4Network):
+        return v
+
+    with change_exception(errors.IPv4NetworkError, ValueError):
+        return IPv4Network(v, strict)
+
+
+def ip_v6_network_validator(v: Any, values: Dict[str, Any]) -> IPv6Network:
+    field: str = 'ip_v6_network_strict'
```
Comment (samuelcolvin) [inline pydantic/validators.py]: and this.

File: pydantic/validators.py [hunk 6]
```diff
@@ -235,6 +235,51 @@ def ip_v6_address_validator(v: Any) -> IPv6Address:
...
+    strict: bool = values.pop(field, True)
+
+    if isinstance(v, IPv4Network):
+        return v
+
+    with change_exception(errors.IPv4NetworkError, ValueError):
+        return IPv4Network(v, strict)
+
+
+def ip_v6_network_validator(v: Any, values: Dict[str, Any]) -> IPv6Network:
```
Comment (samuelcolvin) [inline pydantic/validators.py]: could you add a brief docstring

File: pydantic/types.py (hunk 3 above)
Comment (pilosus) [reply pydantic/types.py]: Ooops, I'm sorry for all that stuff that popped up again! It seems that something went wrong with my rebase. I'll fix that

**Pull request merged**